source/application/DbgHelp/DbgHelpStackWalkCallbackPrivate.cpp

```diff
@@ -0,0 +1,340 @@
+/*
+ * Copyright (c) 2016- Hourglass Resurrection Team
+ * Hourglass Resurrection is licensed under GPL v2.
+ * Refer to the file COPYING.txt in the project root.
+ */
+
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+
+#include <sstream>
+#include <string>
+
+#include <atlbase.h>
+#include <atlcom.h>
+
+#include "application/logging.h"
+#include "DbgHelpPrivate.h"
+#include "DbgHelpStackWalkCallbackPrivate.h"
+
+DbgHelpStackWalkCallbackPrivate::DbgHelpStackWalkCallbackPrivate(HANDLE process, IDiaStackFrame* frame, const DbgHelpPrivate::ModuleData* mod_info) :
+    m_process(process),
+    m_frame(frame),
+    m_mod_info(mod_info),
+    m_unsure(0),
+    m_params_enumerated(false)
+{
+}
+
+// TODO: Split module name and PC
+std::wstring DbgHelpStackWalkCallbackPrivate::GetModuleName()
+{
+    std::wstringstream name;
+    if (m_mod_info == nullptr)
+    {
+        name << L"?";
+    }
+    else
+    {
+        name << m_mod_info->m_module_name;
+    }
+    name << L"!0x" << std::hex << GetProgramCounter();
+    return name.str();
+}
+
+std::wstring DbgHelpStackWalkCallbackPrivate::GetFunctionName()
+{
+    if (m_mod_info == nullptr)
+    {
+        return L"?";
+    }
+    std::wstring function_name = L"?";
+    if (m_mod_info->m_module_symbol_session == nullptr)
+    {
+        auto it = m_mod_info->m_module_exports_table.lower_bound(GetProgramCounter());
+        if (it != m_mod_info->m_module_exports_table.begin())
+        {
+            it--;
+            function_name = it->second;
+        }
+        return function_name;
+    }
+    auto symbol = GetFunctionSymbol();
+    
+    BSTR name = nullptr;
+    if (symbol != nullptr && symbol->get_name(&name) == S_OK)
+    {
+        function_name = name;
+        SysFreeString(name);
+    }
+    return function_name;
+}
+
+// TODO: RFC
+std::wstring DbgHelpStackWalkCallbackPrivate::GetFunctionParameters()

```

Comment:
I suggest some sort of `std::vector<DbgHelp::Parameter>` where `DbgHelp::Parameter` is essentially a tagged union (think `boost::variant`) for the return type.